llvm-external-projects/iree-dialects/lib/Dialect/LinalgExt/IR/LinalgExtOps.cpp

```diff
@@ -2034,6 +2080,202 @@ PackOp::reifyResultShapes(OpBuilder &builder,
+  assert(ivs.size() == getOutputRank() &&
+         "number of ivs must match the rank of the output tensor");
+  OpBuilder::InsertionGuard g(builder);
+  // The `ivs` already represent the position into the output tensor.
+  ReifiedRankedShapedTypeDims outputShape;
+  if (failed(reifyResultShapes(builder, outputShape)))
+    return getOperation()->emitOpError("failed to reify result shape");
+  if (outputShape.size() != 1 || outputShape[0].size() != getOutputRank()) {
+    return getOperation()->emitOpError(
+               "expected shape of one result value of rank")
+           << getOutputRank();
+  }
+
+  DenseMap<int64_t, OpFoldResult> dimAndTileMapping = getDimAndTileMapping();
+  // untiled loops and tile loops induction variables.
+  SmallVector<Value> sourceIvs;
+  // point loops induction variables.
+  SmallVector<Value> sourceIvsPointLoops;
+  sourceIvs.reserve(getOutputRank());
+  sourceIvsPointLoops.reserve(dimAndTileMapping.size());
+  for (auto dim : llvm::seq<int64_t>(0, getOutputRank())) {
+    if (dimAndTileMapping.count(dim)) {
+      DivModValue divMod =
+          getDivMod(builder, loc, ivs[dim],
+                    getAsValues(builder, loc, dimAndTileMapping[dim])[0]);
+      sourceIvsPointLoops.push_back(divMod.remainder);
+      sourceIvs.push_back(divMod.quotient);
+    } else {
+      sourceIvs.push_back(ivs[dim]);
+    }
+  }
+
+  assert(sourceIvsPointLoops.size() + sourceIvs.size() == getInputRank() &&
+         "expect same number of iduction variables equals to input rank");
+  // interchange the point loops induction variables based on `dim_pos`.
+  SmallVector<int64_t> dimsToBlock = extractFromI64ArrayAttr(getDimsPos());
+  SmallVector<int64_t> interchangeVector =
+      computeInterchangeFromDimPos(dimsToBlock, getOutputRank());
+  SmallVector<Value> interchangedSourceIvsPointLoops = sourceIvsPointLoops;
+  interchangedSourceIvsPointLoops = interchange<Value>(
+      interchangedSourceIvsPointLoops, interchangeVector, /*offset=*/0);
+
+  llvm::append_range(sourceIvs, interchangedSourceIvsPointLoops);
+  Value scalar = builder.create<memref::LoadOp>(loc, getInput(), sourceIvs);
+  builder.create<memref::StoreOp>(loc, scalar, getOutput(), ivs);
+  return success();
+}
+
+// TODO: (lorenzo) implement `reifyResultShapes`. Should we also derive
+// `inferResultType` from this method?
+LogicalResult
+UnPackOp::reifyResultShapes(OpBuilder &builder,
+                            ReifiedRankedShapedTypeDims &reifiedReturnShapes) {
+  OpBuilder::InsertionGuard g(builder);
+  builder.setInsertionPoint(getOperation());
+  Location loc = getLoc();
+
+  auto buildOutputDim = [&](OpBuilder &builder, size_t dimIdx) -> OpFoldResult {
+    ArrayRef<int64_t> outputShape = getOutputShape();
+    if (!ShapedType::isDynamic(outputShape[dimIdx])) {
+      return builder.getI64IntegerAttr(outputShape[dimIdx]);
+    }
+    return getDim(builder, loc, getOutput(), dimIdx);
```

Comment:
I don't have full context about it, and we can revisit it later. Considering an e2e test case in IREE, the start IR could be:

```mlir
%init = linalg.init_tensor [...]
%unpack = linalg_ext.unpack %input ... %init ..
```

In this case, we should use input shape to reify the result shape. We had similar issue in pack op, and it's fixed in https://github.com/iree-org/iree/pull/10650

I'd suggest to use getDim(input) + inner_tiles to compute the result shape.

Reply:
If I take it from the input and the tile I need to generate and affine map to compose the two: `#map0 = affine_map<()[s0, s1] -> (s0 + s1)>` where `s0` is the dimesion on the input and `s1` is the tile, which means more IR. Am i miss something?

Reply:
Yes, that means more IR.

After thinking a while, I think the init_tensor carries all the shape information when creating the op. When materialize them into buffers, the information still covers the needs. So they are all workable. We can go either way, and revisit it if there are issues in e2e.